### PR TITLE
Victor/core/adding context

### PR DESCRIPTION
**Pull request by vhvb1989**

### CODE REVIEW

File: sdk/core/core/inc/az_context.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+az_context_with_value(az_context const * parent, void * key, void * value) {
+  return (az_context){
+    ._internal
+    = { .parent = parent, .expiration = _az_context_max_expiration, .key = key, .value = value }
+  };
+}
+
+// Cancels an az_context node in the tree; this effectively cancels all the child nodes as well.
+AZ_INLINE void az_context_cancel(az_context * context) {
+  context->_internal.expiration = 0; // The beginning of time
```
Comment (JeffreyRichter) [inline sdk/core/core/inc/az_context.h]: Insert "if context == NULL) return;" as 1st line in the function

File: sdk/core/core/inc/az_context.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ */
+struct az_context {
+  struct {
+    az_context const * parent; // Pointer to parent context (or NULL); immutable after creation
+    int64_t expiration; // Time when context expires
+    void *key, *value; // Pointers to the key & value (usually NULL)
+  } _internal;
+};
+
+#define _az_context_max_expiration 0x7FFFFFFFFFFFFFFF
```
Comment (antkmsft) [inline sdk/core/core/inc/az_context.h]: I think we should name it as _az_CONTEXT_MAX_EXPIRATION (uppercase)

File: sdk/core/core/inc/az_context.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * @brief An az_context value in a node in a tree that represents expiration times and key/value
+ * pairs. The root node in the tree (ultimate parent) is az_context_app which is a context for the
+ * entire application. Each new node is a child of some parent.
+ */
+struct az_context {
+  struct {
+    az_context const * parent; // Pointer to parent context (or NULL); immutable after creation
+    int64_t expiration; // Time when context expires
+    void *key, *value; // Pointers to the key & value (usually NULL)
```
Comment (antkmsft) [inline sdk/core/core/inc/az_context.h]: Apply the new formatting style, and I also recommend not to use multiple variable declarations.
```
void* key;
void* value;
```

File: sdk/core/core/inc/az_context.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+#define _az_CONTEXT_H
+
+#include <az_result.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+typedef struct az_context az_context;
```
Comment (antkmsft) [inline sdk/core/core/inc/az_context.h]: Can't you define it as `typedef struct { ... } az_context`?

Comment (vhvb1989) [reply sdk/core/core/inc/az_context.h]: no, because the definition of az_context includes az_context inside of it

**Pull request closed without merging**